src/sinks/util/metadata.rs

```diff
@@ -0,0 +1,44 @@
+use super::request_builder::EncodeResult;
+
+/// Metadata for batch requests.
+#[derive(Clone, Debug)]
+pub struct BatchRequestMetadata {
+    /// Number of events represented by this batch request.
+    pub event_count: usize,
+    /// Size, in bytes, of the in-memory representation of all events in this batch request.
+    pub event_byte_size: usize,
+    /// Uncompressed size, in bytes, of the encoded events in this batch request.
+    pub encoded_uncompressed_size: usize,
+    /// Compressed size, in bytes, of the encoded events in this batch request, if compression was performed.
+    pub encoded_compressed_size: Option<usize>,
+}
+
+// TODO: Make this struct the object which emits the actual internal telemetry i.e. events sent, bytes sent, etc.
```

Comment:
My thought is semi-yes.

I'm kind of torn: I think having a type that holds all the metric values be the one to actually emit them is a better separation of concerns than encoding it into `DriverResponse`... but it would, at this point, essentially be a no-op in functionality a lot of refactoring that just inverts from `DriverResponse` to `BatchRequestMetadata`.

This is all further complicated by the notion that, at some point, we should be tracking all bytes sent from a sink -- including retries -- and `DriverResponse` isn't quite the right level to do that, given that it only comes when the request is successful and has no access to retries, etc.

Using `BatchRequestMetadata`, or some additional type that can be passed around as a drop guard, as part of the request itself, would be a natural pathway, and if we're injecting things at that level, having them duplicated at the end with the `DriverResponse` machinery feels like a lot of conceptual overlap.

Long story short, I'd like to have `BatchRequestMetadata` -- probably simply renamed `RequestMetadata` -- be a successor to `DriverResponse` and get extended further. I'm not sure I should explore that design in this PR, though. Hence the TODO.